# What is this Notebook for?
## This Notebook Solely focuses on grouping Data, finding intresting things that raises a doubt. I will take a look on a lot of columns and rows and try to make this exploration fun
### Things that we will focus on :-
- **Finding new and different patterns**
- **Making Series and Dataframes for different tasks**
- **Finding which question is worth asking**

In [119]:
# Lets start by importing pandas
import pandas as pd

In [120]:
# setting maximum columns
pd.set_option('display.max_columns', 100)

In [121]:
# lets import the filtered data that we made in 02_data_cleaning.ipynb
df60 = pd.read_csv('../Data/Filtered/ucr_1960_cleaned.csv')
df61 = pd.read_csv('../Data/Filtered/ucr_1961_cleaned.csv')
df62 = pd.read_csv('../Data/Filtered/ucr_1962_cleaned.csv')
df = pd.read_csv('../Data/Filtered/ucr_1960_1962_cleaned.csv')

In [122]:
df.head()

,ST,ORI,GRP,DIV,YEAR,MSA,CD1,CD2,CD3,ADJD,OFF,M3,M4,M5,M6,M8,M9,M10,M11,M12,M13,M14,M15,M16,M17,M18,M19,M20,M24,F3,F4,F5,F6,F8,F9,F10,F11,F12,F13,F14,F15,F16,F17,F18,F19,F20,F24,AW,AB,AI,AO1,AH,AN
0,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",50,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0
1,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",60,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0
2,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",150,0,0,1,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0
3,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",210,0,0,1,0,0,0,0,1,0,0,0,6,4,2,2,2,9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,29,2,0,0,0,0
4,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,0,1,2,0,0,0,0,0,2,0,1,7,5,9,13,3,23,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,0,63,9,0,0,0,0


In [123]:
# checking which crime code has been mentioned the most
df.OFF.value_counts()

OFF
60     4216
50     4164
240    4146
210    4100
230    4044
260    3826
70     3735
80     3686
170    3224
150    2832
220    2790
40     2762
250    2693
200    2667
110    2553
100    2454
30     2288
130    2072
190    1755
20     1753
270    1739
180    1527
11     1376
12     1246
160     838
Name: count, dtype: int64

In [124]:
# checking which crime code has been mentioned the least
df.OFF.value_counts(ascending=True)

OFF
160     838
12     1246
11     1376
180    1527
270    1739
20     1753
190    1755
130    2072
30     2288
100    2454
110    2553
200    2667
250    2693
40     2762
220    2790
150    2832
170    3224
80     3686
70     3735
260    3826
230    4044
210    4100
240    4146
50     4164
60     4216
Name: count, dtype: int64

#### First question that we will ask
**which type of crime had the most and least reports over the three year and each year specifically?**

In [125]:
# checking the shape and d-type of each
print('Shapes')
print(f'Shape comprising all three years is :{df.shape}')
print(f'Shape comprising year 1960 is :{df60.shape}')
print(f'Shape comprising year 1961 is :{df61.shape}')
print(f'Shape comprising year 1962 is :{df62.shape}')
print()
print('D-types')
print('As columns are same for all, d-types is also same')
print(f'D-Type for all columns are :{df.dtypes}')

Shapes
Shape comprising all three years is :(68486, 53)
Shape comprising year 1960 is :(20558, 53)
Shape comprising year 1961 is :(23232, 53)
Shape comprising year 1962 is :(24696, 53)

D-types
As columns are same for all, d-types is also same
D-Type for all columns are :ST        str
ORI       str
GRP       str
DIV       str
YEAR    int64
MSA     int64
CD1       str
CD2       str
CD3       str
ADJD      str
OFF     int64
M3      int64
M4      int64
M5      int64
M6      int64
M8      int64
M9      int64
M10     int64
M11     int64
M12     int64
M13     int64
M14     int64
M15     int64
M16     int64
M17     int64
M18     int64
M19     int64
M20     int64
M24     int64
F3      int64
F4      int64
F5      int64
F6      int64
F8      int64
F9      int64
F10     int64
F11     int64
F12     int64
F13     int64
F14     int64
F15     int64
F16     int64
F17     int64
F18     int64
F19     int64
F20     int64
F24     int64
AW      int64
AB      int64
AI      int64
AO1     int64
AH      int64


In [126]:
# Checking Maximum Crime by State
df.ST.value_counts().head()

ST
California      9593
New Jersey      5782
Pennsylvania    5057
Illinois        4812
New York        4741
Name: count, dtype: int64

In [127]:
# Checking Minimum Crime by State
df.ST.value_counts(ascending=True).head()

ST
South Dakota.    55
Mississippi      69
Hawaii           72
New Mexico       75
Delaware         98
Name: count, dtype: int64

#### A question can be asked from this
**Which state had the most crime for each year and overall?**

In [128]:
# Lets check the total arrest by gender and different categories
AGE_M = [f'M{i}' for i in range(3, 25) if i not in (7, 21, 22, 23)]   # M3..M24
AGE_F = [f'F{i}' for i in range(3, 25) if i not in (7, 21, 22, 23)]   # F3..F24
df['total_male']    = df[AGE_M].sum(axis=1)
df['total_female']  = df[AGE_F].sum(axis=1)
df['total_arrests'] = df['total_male'] + df['total_female']

In [129]:
print(f"Total male arrests:   {df['total_male'].sum():,}")
print(f"Total female arrests: {df['total_female'].sum():,}")
print(f"Total arrests:        {df['total_arrests'].sum():,}")

Total male arrests:   9,170,166
Total female arrests: 1,206,565
Total arrests:        10,376,731


In [130]:
# Per-year breakdown
print("\nBy year:")
print(df.groupby('YEAR')[['total_male', 'total_female', 'total_arrests']].sum())


By year:
      total_male  total_female  total_arrests
YEAR                                         
1960     3003066        384670        3387736
1961     2987107        395523        3382630
1962     3179993        426372        3606365


In [131]:
yearly = df.groupby('YEAR').agg(
    rows=('ORI', 'size'),
    agencies=('ORI', 'nunique'),
    total=('total_arrests', 'sum'),
)
yearly['arrests_per_agency'] = yearly['total'] / yearly['agencies']
print(yearly)

       rows  agencies    total  arrests_per_agency
YEAR                                              
1960  20558      1317  3387736         2572.312832
1961  23232      1554  3382630         2176.724582
1962  24696      1664  3606365         2167.286659


## Observation: raw year totals grow, but rows grow faster

- 1960: 3,387,736 arrests across 20,558 rows
- 1961: 3,382,630 arrests across 23,232 rows
- 1962: 3,606,365 arrests across 24,696 rows

Raw totals look roughly stable. But the number of reporting agencies grew
by ~20% over the same period.

**Question for final analysis:** Did arrests per reporting agency change
from 1960 to 1962? Did the *composition* of arrests shift, rather than
the absolute count?

In [132]:
# Groupby State
df.groupby('ST')[['total_male', 'total_female', 'total_arrests']].sum()

,total_male,total_female,total_arrests
ST,,,
Alabama,92920,13106,106026
Arizona,111021,12524,123545
Arkansas,34469,4547,39016
California,1558289,190242,1748531
Colorado,115657,15417,131074
Connecticut,100934,11820,112754
Delaware,31700,4103,35803
District of Columbia,254797,34412,289209
Florida,229610,38529,268139


In [133]:
# checking the state with the highest and lowest total arrests in different categories
df.groupby('ST')[['total_male', 'total_female', 'total_arrests']].sum().agg(['idxmax', 'idxmin', 'max', 'min'])

,total_male,total_female,total_arrests
idxmax,California,California,California
idxmin,South Dakota.,South Dakota.,South Dakota.
max,1558289,190242,1748531
min,4071,459,4530


## Questions for Final Analysis

**Quick facts:**
- Total male vs female arrests, 1960–1962?
- Total arrests per year — which year highest, which lowest?
- State with the most and fewest arrests overall?

**Worth a real analysis:**
- Did arrests per reporting agency change 1960 → 1962? (Raw totals grew, but so did agencies reporting.)
- Which age bracket has the highest arrest rate per year-of-age? (Raw counts favor 5-year brackets.)
- What share of arrests were public-order offenses (210, 220, 230, 240, 250)? Does it vary by year or region?
- Do the largest agencies have a different offense profile than the rest?
- Which offenses show the widest male/female split?
- Does offense composition differ by census division?

**Notes for the record:**
- Zero-arrest rows are rare (0.19%) — file is arrest-centric, not agency-centric.
- Juvenile race data (J* columns) is empty in all three years.
- 1961 shows a small variance in reporting flags (`CD1`–`CD3`, `ADJD`) that 1960/1962 don't.


In [134]:
df.head(10)

,ST,ORI,GRP,DIV,YEAR,MSA,CD1,CD2,CD3,ADJD,OFF,M3,M4,M5,M6,M8,M9,M10,M11,M12,M13,M14,M15,M16,M17,M18,M19,M20,M24,F3,F4,F5,F6,F8,F9,F10,F11,F12,F13,F14,F15,F16,F17,F18,F19,F20,F24,AW,AB,AI,AO1,AH,AN,total_male,total_female,total_arrests
0,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",50,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,1
1,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",60,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,1
2,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",150,0,0,1,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0,3,0,3
3,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",210,0,0,1,0,0,0,0,1,0,0,0,6,4,2,2,2,9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,29,2,0,0,0,0,27,2,29
4,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,0,1,2,0,0,0,0,0,2,0,1,7,5,9,13,3,23,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,0,63,9,0,0,0,0,66,6,72
5,Alabama,AL00103,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",240,0,0,2,1,0,0,0,0,0,0,0,1,1,0,1,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6,2,0,0,0,0,8,0,8
6,Alabama,AL00104,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",40,0,0,0,2,2,4,2,0,1,0,0,2,1,1,3,3,4,0,0,0,0,0,0,0,0,0,1,0,0,1,1,1,0,0,1,0,5,25,0,0,0,0,25,5,30
7,Alabama,AL00104,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",60,0,0,6,10,4,1,4,0,0,0,0,3,2,6,0,2,0,0,0,0,0,1,1,0,0,1,0,0,0,2,2,1,0,1,0,0,9,38,0,0,0,0,38,9,47
8,Alabama,AL00104,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",80,0,0,0,1,3,1,2,0,1,1,2,7,10,4,5,4,4,0,0,0,1,0,1,0,1,0,0,1,0,1,2,2,0,1,2,0,12,45,0,0,0,0,45,12,57
9,Alabama,AL00104,5,East South Central,1960,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",100,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,1,0,0,0,0,1,1,2


In [135]:
# which row had the most no of arrests
df.sort_values('total_arrests', ascending=False).head(20)

,ST,ORI,GRP,DIV,YEAR,MSA,CD1,CD2,CD3,ADJD,OFF,M3,M4,M5,M6,M8,M9,M10,M11,M12,M13,M14,M15,M16,M17,M18,M19,M20,M24,F3,F4,F5,F6,F8,F9,F10,F11,F12,F13,F14,F15,F16,F17,F18,F19,F20,F24,AW,AB,AI,AO1,AH,AN,total_male,total_female,total_arrests
1402,California,CA01942,1A,Pacific,1960,478,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,47,71,144,257,530,588,602,910,971,897,997,6260,9342,11825,13336,11548,25387,0,12,13,11,25,33,29,39,79,131,137,111,712,1011,1149,1432,1001,1368,0,70044,17148,3463,24,217,109,83712,7293,91005
22128,California,CA01942,1A,Pacific,1961,478,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,93,114,169,279,606,599,709,1013,981,1019,932,5892,8413,10974,12541,11020,24970,0,22,12,17,16,50,50,49,89,111,116,100,635,816,1159,1330,855,1430,0,66704,17370,2732,27,208,140,80324,6857,87181
45509,California,CA01942,1A,Pacific,1962,478,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,75,126,187,270,568,618,657,968,1036,970,960,5576,7876,10642,13106,10884,25103,0,22,25,31,34,38,79,63,103,117,118,118,602,845,1056,1239,770,1298,0,66347,16437,3059,22,192,123,79622,6558,86180
49639,Illinois,ILCPD00,1A,East North Central,1962,161,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,320,43,77,387,479,566,503,945,1008,929,942,5919,9306,12384,12456,9465,17243,0,46,23,9,10,36,47,57,88,87,74,98,389,474,501,423,328,432,0,45128,25157,2748,9,4,3048,72972,3122,76094
60581,New York,NY03030,1A,Middle Atlantic,1962,607,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",240,282,369,954,1034,994,876,671,1392,963,945,1037,5488,7799,9577,10557,8492,19704,0,31,22,40,31,50,55,42,207,88,85,111,559,587,325,198,127,139,0,0,0,0,0,0,0,71134,2697,73831
35982,New York,NY03030,1A,Middle Atlantic,1961,607,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,No Race Reported,240,278,319,752,874,776,614,560,987,677,662,673,3972,6120,7310,7798,7157,16136,0,17,15,29,31,40,33,28,180,96,126,145,696,499,295,146,105,128,0,0,0,0,0,0,0,55665,2609,58274
4757,Illinois,ILCPD00,1A,East North Central,1960,161,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,0,0,0,86,110,99,110,383,397,399,420,3007,6437,9810,8893,8180,14888,0,0,0,0,0,5,14,16,12,20,15,15,88,137,187,147,179,193,0,42443,8823,2933,1,9,38,53219,1028,54247
4557,Georgia,GAAPD00,1C,South Atlantic,1960,50,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,8,6,30,81,151,226,259,397,470,605,638,4261,5973,7632,7333,6618,10813,0,6,2,6,19,56,48,67,95,124,139,149,1074,1321,1568,1128,919,958,0,26353,26827,0,0,0,0,45501,7679,53180
13624,New York,NY03030,1A,Middle Atlantic,1960,607,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",240,247,351,769,958,700,565,484,866,561,524,485,3150,4929,6498,6536,6504,14314,0,18,17,7,19,23,18,19,134,53,56,62,315,245,181,121,68,94,0,0,0,0,0,0,0,48441,1450,49891
25739,Georgia,GAAPD00,1C,South Atlantic,1961,50,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,6,7,30,129,201,289,267,529,567,678,625,4470,5752,7019,6646,5437,9888,0,3,5,4,39,44,61,78,114,128,131,134,990,1291,1457,1072,718,1058,0,22322,27545,0,0,0,0,42540,7327,49867


In [136]:
# which row had the least no of arrests
df.sort_values('total_arrests', ascending=True)

,ST,ORI,GRP,DIV,YEAR,MSA,CD1,CD2,CD3,ADJD,OFF,M3,M4,M5,M6,M8,M9,M10,M11,M12,M13,M14,M15,M16,M17,M18,M19,M20,M24,F3,F4,F5,F6,F8,F9,F10,F11,F12,F13,F14,F15,F16,F17,F18,F19,F20,F24,AW,AB,AI,AO1,AH,AN,total_male,total_female,total_arrests
89,Alabama,AL00300,9B,East South Central,1960,556,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",70,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0
8775,Massachusetts,MA01416,5,New England,1960,979,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",220,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0
8882,Massachusetts,MA01454,5,New England,1960,979,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",270,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0
8599,Massachusetts,MA01220,5,New England,1960,115,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",30,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0
56666,Minnesota,MN06908,6,West North Central,1962,234,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",260,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
60581,New York,NY03030,1A,Middle Atlantic,1962,607,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",240,282,369,954,1034,994,876,671,1392,963,945,1037,5488,7799,9577,10557,8492,19704,0,31,22,40,31,50,55,42,207,88,85,111,559,587,325,198,127,139,0,0,0,0,0,0,0,71134,2697,73831
49639,Illinois,ILCPD00,1A,East North Central,1962,161,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,320,43,77,387,479,566,503,945,1008,929,942,5919,9306,12384,12456,9465,17243,0,46,23,9,10,36,47,57,88,87,74,98,389,474,501,423,328,432,0,45128,25157,2748,9,4,3048,72972,3122,76094
45509,California,CA01942,1A,Pacific,1962,478,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,75,126,187,270,568,618,657,968,1036,970,960,5576,7876,10642,13106,10884,25103,0,22,25,31,34,38,79,63,103,117,118,118,602,845,1056,1239,770,1298,0,66347,16437,3059,22,192,123,79622,6558,86180
22128,California,CA01942,1A,Pacific,1961,478,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,93,114,169,279,606,599,709,1013,981,1019,932,5892,8413,10974,12541,11020,24970,0,22,12,17,16,50,50,49,89,111,116,100,635,816,1159,1330,855,1430,0,66704,17370,2732,27,208,140,80324,6857,87181


In [137]:
#lets check how many rows has 0 total arrests
(df['total_arrests'] == 0).sum()

np.int64(133)

Only 133 rows (0.19%) have total_arrests == 0.

That's much lower than you'd expect if the file listed every agency × offense
combination, including zeros. It suggests the detail file only includes rows
where the agency had at least one arrest to report — the zeros we see are
likely edge cases or records filed for other reasons.

**Question for final analysis:** Does this mean absent rows should be treated
as zero-arrest cases? If so, agency-level averages might be biased toward
agencies that arrest more often. Worth checking whether the 133 zero-rows
cluster in a specific state, year, or offense.

## Observation: public-order offenses dominate the top of the code list

Top 2 offense codes by row count:240, 230.

of those (230 Drunkenness, 240 Disorderly Conduct) are
public-order offenses, not the "headline" crimes (murder, robbery, assault).

**Question for final analysis:** What share of total arrests were
public-order offenses, and did that share change across years or regions?

In [138]:
# Age brackets aren't the same width.
#
# M4 is a single year of age (15), but M15 covers five years (25-29).
# A bracket that covers 5 years will naturally collect more arrests than
# one that covers 1 year — not because that age group commits more crime,
# just because there are more birth years inside it.
#
# To compare fairly, divide each bracket's total by how many years it spans.
# That gives arrests per year-of-age, so every bracket is on the same footing.

BRACKET_WIDTH = {
    '3': 2,   # 13-14
    '4': 1,   # 15
    '5': 1,   # 16
    '6': 1,   # 17
    '8': 1,   # 18
    '9': 1,   # 19
    '10': 1,  # 20
    '11': 1,  # 21
    '12': 1,  # 22
    '13': 1,  # 23
    '14': 1,  # 24
    '15': 5,  # 25-29
    '16': 5,  # 30-34
    '17': 5,  # 35-39
    '18': 5,  # 40-44
    '19': 5,  # 45-49
    '20': 5,  # 50-54
    '24': 1,  # no age (adult) — treated as a single bucket
}

age_totals = df[AGE_M].sum()

age_per_year = pd.Series({
    col: age_totals[col] / BRACKET_WIDTH[col[1:]]
    for col in AGE_M
})

age_per_year.sort_values(ascending=False)

M20    307028.6
M6     287753.0
M3     281476.5
M8     276677.0
M5     275336.0
M9     255366.0
M11    245119.0
M4     234533.0
M10    227815.0
M12    223406.0
M13    206474.0
M17    204357.4
M14    201087.0
M16    195056.2
M15    185068.2
M18    184568.8
M19    157604.0
M24      5231.0
dtype: float64

In [139]:
# Age brackets aren't the same width.
#
# F4 is a single year of age (15), but F15 covers five years (25-29).
# A bracket that covers 5 years will naturally collect more arrests than
# one that covers 1 year — not because that age group commits more crime,
# just because there are more birth years inside it.
#
# To compare fairly, divide each bracket's total by how many years it spans.
# That gives arrests per year-of-age, so every bracket is on the same footing.

BRACKET_WIDTH = {
    '3': 2,   # 13-14
    '4': 1,   # 15
    '5': 1,   # 16
    '6': 1,   # 17
    '8': 1,   # 18
    '9': 1,   # 19
    '10': 1,  # 20
    '11': 1,  # 21
    '12': 1,  # 22
    '13': 1,  # 23
    '14': 1,  # 24
    '15': 5,  # 25-29
    '16': 5,  # 30-34
    '17': 5,  # 35-39
    '18': 5,  # 40-44
    '19': 5,  # 45-49
    '20': 5,  # 50-54
    '24': 1,  # no age (adult) — treated as a single bucket
}

age_totals = df[AGE_F].sum()

age_per_year = pd.Series({
    col: age_totals[col] / BRACKET_WIDTH[col[1:]]
    for col in AGE_F
})

age_per_year.sort_values(ascending=False)

F3     56451.0
F4     51871.0
F5     47781.0
F6     35139.0
F12    33721.0
F11    32934.0
F13    32304.0
F14    31480.0
F16    29521.2
F17    29229.2
F15    29092.4
F9     28895.0
F8     28698.0
F10    26268.0
F18    22804.6
F20    22120.2
F19    15890.8
F24     1280.0
dtype: float64

## Observation: age bracket ranking looks like a bracket-width artifact

Raw totals rank M20 (50-54) at the top with 1,531,143 arrests, above M4
(15), M5 (16), etc.

But the brackets aren't the same width. M20 covers 5 years of age; M4
covers 1. Without normalizing, we're comparing "arrests per 5-year bracket"
against "arrests per 1-year bracket" — which isn't fair.

**Question for final analysis:** Which age bracket has the highest
arrest rate *per group of age*? Does the ranking change?

In [140]:
# Which agencies reported the most total arrests?
# Sanity check: should be the big-city departments we'd expect, not tiny towns.
df.groupby(['ORI', 'ST', 'GRP'])['total_arrests'].sum().nlargest(10)

ORI      ST                    GRP
NY03030  New York              1A     537182
CA01942  California            1A     523268
ILCPD00  Illinois              1A     389505
PAPEP00  Pennsylvania          1A     388915
DCMPD00  District of Columbia  1B     284562
TXDPD00  Texas                 1B     283232
LANPD00  Louisiana             1B     189248
MOSPD00  Missouri              1B     185330
GAAPD00  Georgia               1C     181077
MI82349  Michigan              1A     150098
Name: total_arrests, dtype: int64

**Question for final analysis:** Do the largest agencies have a different
offense profile than the national average? Specifically, do big-city
departments arrest more people for public-order offenses (drunkenness,
disorderly conduct) relative to serious crimes?

In [141]:
# Public-order share of arrests, by year.
# Codes: 210 DUI, 220 liquor, 230 drunkenness, 240 disorderly, 250 vagrancy.
public_order = [210, 220, 230, 240, 250]

by_year = df.groupby('YEAR').apply(
    lambda g: g.loc[g['OFF'].isin(public_order), 'total_arrests'].sum()
              / g['total_arrests'].sum()
)
print(by_year)

YEAR
1960    0.563923
1961    0.555070
1962    0.555855
dtype: float64


## Thread: public-order arrest share

More than half of all arrests in every year (55–56%) were public-order
offenses (210, 220, 230, 240, 250). Stable across 1960, 1961, 1962.

Questions worth asking in final analysis:

- Does the share vary by census division?
- Does it vary by agency size (GRP)?
- Which specific code drives the share — is it drunkenness alone, or spread?
- Do big-city agencies (top 10 by total arrests) differ from the national average?

Possible final research question that can add:

> Does the public-order share of arrests vary by region and agency size,
> and is it stable across 1960–1962?

In [142]:
# Which offenses show the widest male/female split?
# A low female_share means male-dominated; a high one means near parity.
top10 = df.groupby('OFF')['total_arrests'].sum().nlargest(10).index
gender = df.groupby('OFF')[['total_male', 'total_female']].sum()
gender['female_share'] = gender['total_female'] / (gender['total_male'] + gender['total_female'])

print(gender.loc[top10].sort_values('female_share'))

     total_male  total_female  female_share
OFF                                        
50       350752         12722      0.035001
210      390672         28200      0.067324
230     3274555        298963      0.083661
190      306598         28301      0.084506
250      352414         32534      0.084515
80       347269         40980      0.105551
270      277477         36541      0.116366
240     1018052        166870      0.140828
260     1137574        226349      0.165954
60       501855        115439      0.187008


**Questions for final analysis:**
- Which offense categories have the widest gender gap?
- Is the gender ratio stable across 1960–1962?
- Does the male/female split vary by region or agency size for the same offense?

## Wrapping up EDA - moving to final analysis

Okay so this notebook did its job. We poked around the data, found some weird stuff, and now i have a much better idea of what questions are actually worth asking. Before i start the final analysis notebook, just wanna write down what i found and what the limits are.

### Main things i found

**1. The reporting artifact (this one is important)**
Raw total arrests go up from 1960 to 1962, but thats super misleading. The number of agencies reporting also went up by like 20% in the same time. So we cant just say "crime went up". For the final analysis i need to either look at arrests_per_agency, or only use the agencies that reported in all 3 years. Otherwise the numbers are just lying to us.

**2. Public order stuff is basically half the data**
Drunkenness (230) and disorderly conduct (240) are always at the top. More than half of ALL arrests (55-56%) every single year were public order offenses. And that share barely moved between 60, 61 and 62 which is kinda interesting.

**3. Huge gender gap**
Males get arrested way more than females, like not even close. But the gap changes depending on the offense. For public order and vagrancy stuff its mostly dudes. For larceny (60) the female share is a little higher.

**4. The age bracket thing was a trap**
At first i thought M20 (50-54) had the most arrests. But then i realized M20 covers 5 years of age and M4 only covers 1 year. So of course the bigger bracket has more arrests lol. Once i divided by the number of years in each bracket, the ranking totally changed and younger brackets showed up at the top. Good catch.

### Stuff to watch out for (data limitations)

* The data is arrest-centric not agency-centric. Only 0.19% of rows are zero arrests. So if a row is missing it probably means the agency just didnt report it, not that there were zero arrests. Gotta be careful with agency averages.
* 1961 has some small differences in the reporting flags (CD1-CD3, ADJD) compared to 60 and 62. Not a huge deal but worth remembering.
* No population data in here. So i cant do per-capita rates. Everything is raw counts or per agency.

### What im doing next (final analysis questions)

1. Answering important questions, that we marked in this file.
2. A detailed explanation of what the data means in actual form
3. checking if we left any analysis category
4. Making graphs and plots of the data to make it visually more appealing

Alright time to open a new notebook and actually answer some of these.